File: spring-web-java/src/main/java/spring/web/java/domain/model/entity/Member.java
package spring.web.java.domain.model.entity;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import org.springframework.util.StringUtils;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Convert;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.Id;
import jakarta.persistence.OneToMany;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import spring.web.java.domain.converter.CryptoAttributeConverter;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Member extends Base {

	@Id
	@GeneratedValue
	@Column(name = "member_id")
	private Long id;
	private String account;
	private String password;

	@Convert(converter = CryptoAttributeConverter.class)
	private String name;

	@Convert(converter = CryptoAttributeConverter.class)
	private String phone;

	private LocalDate birthDate;

	@OneToMany(mappedBy = "member", cascade = CascadeType.ALL)
	private List<MemberRole> memberRoles = new ArrayList<>();

	@OneToMany(mappedBy = "member", cascade = CascadeType.ALL)
	private List<MemberPermission> memberPermissions = new ArrayList<>();

	public List<MemberRole> getMemberRoles() {
		return Collections.unmodifiableList(memberRoles);
	}

	public List<MemberPermission> getMemberPermissions() {
		return Collections.unmodifiableList(memberPermissions);
	}

	public static Member create(
		String account,
		String password,
		String name,
		List<MemberRole> memberRoles,
		List<MemberPermission> memberPermissions
	) {
		Member member = new Member();

		member.account = account;
		member.password = password;
		member.name = name;

		memberRoles.forEach(member::associateRole);
		memberPermissions.forEach(member::associatePermission);

		return member;
	}

	public void associateRole(MemberRole memberRole) {
		memberRoles.add(memberRole);
		memberRole.associateMember(this);
	}

	public void associatePermission(MemberPermission memberPermission) {
		memberPermissions.add(memberPermission);
		memberPermission.associateMember(this);
	}

	public void update(String password, String name, String phone, LocalDate birthDate) {
		if (StringUtils.hasText(password)) {
			this.password = password;
		}

		if (StringUtils.hasText(name)) {
			this.name = name;
		}

		if (StringUtils.hasText(phone)) {
			this.phone = phone;
		}

		if (birthDate != null) {
			this.birthDate = birthDate;
		}
	}
}

File: spring-web-kotlin/src/main/kotlin/spring/web/kotlin/domain/model/entity/Member.kt
package spring.web.kotlin.domain.model.entity

import jakarta.persistence.*
import spring.web.kotlin.domain.converter.CryptoAttributeConverter
import java.time.LocalDate

@Entity
class Member protected constructor(
    @Convert(converter = CryptoAttributeConverter::class)
    val account: String,
    password: String,
    name: String,
    phone: String,
    birthDate: LocalDate,
) : BaseTime() {
    @Id
    @GeneratedValue
    @Column(name = "member_id")
    var id: Long? = null
        protected set

    var password: String = password
        protected set

    @Convert(converter = CryptoAttributeConverter::class)
    var name: String = name
        protected set

    @Convert(converter = CryptoAttributeConverter::class)
    var phone: String = phone
        protected set

    var birthDate: LocalDate = birthDate
        protected set

    @OneToMany(mappedBy = "member", cascade = [(CascadeType.ALL)])
    private val _memberRoles: MutableList<MemberRole> = mutableListOf()

    @get:Transient
    val memberRoles: List<MemberRole>
        get() = _memberRoles.toList()

    @OneToMany(mappedBy = "member", cascade = [(CascadeType.ALL)])
    private val _memberPermissions: MutableList<MemberPermission> = mutableListOf()

    @get:Transient
    val memberPermissions: List<MemberPermission>
        get() = _memberPermissions.toList()

    companion object {
        fun create(
            account: String,
            password: String,
            name: String,
            phone: String,
            birthDate: LocalDate,
            memberRoles: List<MemberRole>,
            memberPermissions: List<MemberPermission>
        ) = Member(
            account = account,
            password = password,
            name = name,
            phone = phone,
            birthDate = birthDate,
        ).apply {
            memberRoles.forEach { this.associateRole(it) }
            memberPermissions.forEach { this.associatePermission(it) }
        }
    }

    fun associateRole(memberRole: MemberRole) {
        _memberRoles.add(memberRole)
        memberRole.associateMember(this)
    }

    fun associatePermission(memberPermission: MemberPermission) {
        _memberPermissions.add(memberPermission)
        memberPermission.associateMember(this)
    }

    fun update(password: String?, name: String?, phone: String?, birthDate: LocalDate?) {
        password?.let { this.password = password }
        name?.let { this.name = name }
        phone?.let { this.phone = phone }
        birthDate?.let { this.birthDate = birthDate }
    }
}
